### DISCUSSION TITLE
How to Debug Native Compile Times

### DISCUSSION THREAD

**Discussion by justinpapezsonos** [android/ndk#1703] [Q&A] [answered]
How should one go about debugging / diagnosing / resolving exuberant native compile times? The company that I work for recently migrated from NDK r21 to NDK r24 and some developers have seen their compile times increase on average from 4 minutes to 1 hour. These developers have noticed that the compile time regression occurs when building with NDK r23+ (they have tested with NDK r21, r22, r23, r24, & r25). The native library being compiled is roughly 1,200 files. Unfortunately, we are not sure how to diagnose what the root issue may be. We'd like to use NDK r23+ as it supports M1 machines, something that is important to our company. Any suggestions or help as to how we can proceed in order to diagnose the issue would be greatly appreciated. Thanks in advance!

**Comment by pirama-arumuga-nainar** [comment, marked as answer]
tl;dr: Antivirus was slowing things down. See the full thread for details.

A jump from a few minutes to an hour seems surprising.  There was no
noticeable slowdown in the Android platform or kernel build times.

Does anything stand out when observing build progress?  For e.g., is the
slowdown uniform across all sources or is it just a few files that are
bottlenecked?

**Reply by justinpapezsonos** [reply]
Hey @pirama-arumuga-nainar - I agree, it's extraordinarily surprising, we are equally shocked. The build time for some has increased almost 20x - I am not exaggerating. We have not been able to isolate what a potential cause may be yet. We are building across Linux machines, Intel based Macs, and Silicon based Macs. So far, we have only noticed the increased build times on Macs, Linux seems to be unaffected.

The slowdown is uniform across all source files for those who are experiencing significantly increased build times. It is not bottlenecked to a few files as far as we can tell.

What we can say for certain is that downgrading from NDK r24 to r22 immediately stabilizes build times for our developers. The moment we move back to NDK r23 or higher, some developers begin experiencing increased build times again.

**Reply by pirama-arumuga-nainar** [reply]
IIUC, r23 has the clang toolchain as universal binaries (https://github.com/android/ndk/issues/1544).  The slowdown may be connected to that.

> some developers begin experiencing increased build times again

Any commonalities across these MacOS machines?  @kongy: do you have any idea on slowdowns when switching to universal binaries?

**Reply by kongy** [reply]
No, universal binary is not supposed to cause significant slowdown. One way to debug this is to reduce universal binary to regular binary running `lipo -extract x86_64 clang-13 -o ...` and observe the performance difference.

**Reply by DanAlbert** [reply]
We expect the universal binaries to cause a significant performance _improvement_ for M1. Unless the machine is for some reason spending a ton more time loading the binaries from disk than it is actually compiling, that seems unlikely.

If the slow down is noticeable outside a full build (`clang++ $FLAGS $SOME_SOURCE_FILE` is significantly slower alone), strace might tell you something interesting. Penalties that severe make me think it could be something gatekeeper/AV/audit related.

**Reply by enh-google** [reply]
s/strace/dtruss/ on macOS :-) (or "Instruments" if you're in to the whole GUI thing...)

**Reply by enh-google** [reply]
one question i'm not sure has been explicitly answered is whether they're seeing this on x86-64 macs, M1 macs, or both?

**Reply by justinpapezsonos** [reply]
Thanks for your quick replies!

> IIUC, r23 has the clang toolchain as universal binaries (#1544). The slowdown may be connected to that.
> 
> > some developers begin experiencing increased build times again
> 
> Any commonalities across these MacOS machines? @kongy: do you have any idea on slowdowns when switching to universal binaries?

No commonalities that we can discern as of yet. 

> No, universal binary is not supposed to cause significant slowdown. One way to debug this is to reduce universal binary to regular binary running `lipo -extract x86_64 clang-13 -o ...` and observe the performance difference.

I'll coordinate with some of the folks experiencing the increased build times and see if they can run a test against this and then report back.

> We expect the universal binaries to cause a significant performance _improvement_ for M1. Unless the machine is for some reason spending a ton more time loading the binaries from disk than it is actually compiling, that seems unlikely.
> 
> If the slow down is noticeable outside a full build (`clang++ $FLAGS $SOME_SOURCE_FILE` is significantly slower alone), strace might tell you something interesting. Penalties that severe make me think it could be something gatekeeper/AV/audit related.

I'll coordinate with some of the folks experiencing the increased build times and see if they can run a test against this and then report back.

> one question i'm not sure has been explicitly answered is whether they're seeing this on x86-64 macs, M1 macs, or both?

Here are a few of the machine specs. we have seen slowdowns on:
- 2019 Intel MBP, 32GB RAM, 1TB
- 2019 Intel MBP, 32GB RAM, 2TB
- M1 Macbook Air, 16GB RAM, 512 GB

**Reply by justinpapezsonos** [reply]
Just wanted to share a few quick updates:

> IIUC, r23 has the clang toolchain as universal binaries (#1544). The slowdown may be connected to that.
> 
> > some developers begin experiencing increased build times again
> 
> Any commonalities across these MacOS machines? @kongy: do you have any idea on slowdowns when switching to universal binaries?

While not all developers at my company are experiencing the aforementioned increased build times, those who are have noticed that there seems to be no parallelization when compiling our native library (which is unexpected). Specifically, there are ~14 clang processes running, but only one of them seems to be under load. Please see the following screenshots of the clang processes in Activity Monitor and a general view of CPU History when our native library is being compiled.

<img width="1459" alt="Screen Shot 2022-04-21 at 8 44 02 AM" src="https://user-images.githubusercontent.com/42981091/164759906-6b86b747-73f1-43ab-93f5-2551d67884fa.png">
<img width="658" alt="Screen Shot 2022-04-21 at 8 34 52 AM" src="https://user-images.githubusercontent.com/42981091/164790701-559180f5-99ad-45d9-8371-3550c9107533.png">

When these developers revert back to NDK r21, they immediately see all ~14 clang processes under load and much higher core utilization via CPU History. Please see the following screenshot.

<img width="658" alt="Screen Shot 2022-04-21 at 9 02 48 AM" src="https://user-images.githubusercontent.com/42981091/164759995-cfd0f2e3-93aa-433d-8628-25414b0d3018.png">

We are using Ninja / CMake for what it's worth.

> No, universal binary is not supposed to cause significant slowdown. One way to debug this is to reduce universal binary to regular binary running `lipo -extract x86_64 clang-13 -o ...` and observe the performance difference.

We stripped the universal binary into regular binaries and observed no difference in performance (which it sounds like is expected based on comments from kongy & Dan Albert).

> We expect the universal binaries to cause a significant performance _improvement_ for M1. Unless the machine is for some reason spending a ton more time loading the binaries from disk than it is actually compiling, that seems unlikely.
> 
> If the slow down is noticeable outside a full build (`clang++ $FLAGS $SOME_SOURCE_FILE` is significantly slower alone), strace might tell you something interesting. Penalties that severe make me think it could be something gatekeeper/AV/audit related.

Still coordinating with a few folks to have them run some tests via strace.

**Reply by DanAlbert** [reply]
Which r21? If you go back far enough (r21b or older) the binaries won't be signed/notarized. If gatekeeper is for some reason blocking all those processes only when they're signed/notarized, that's a difference. I am inclined to blame gatekeeper here, since this does look very similar to the behavior you'd find on Windows when a process is blocked waiting on AV.

Are all your developers downloading the NDK the same way? Via the SDK manager, or via direct download? For direct downloads there are two different NDKs available for Mac: an app bundle (Apple requires the NDK be distributed this way for notarization, which gatekeeper requires) and the normal zip file, which cannot be notarized. The SDK can only distribute the zip file, but because the SDK manager doesn't quarantine its downloads those binaries don't require notarization. Maybe the behavior split is based on the download type/method?

Another possible split would be the settings on each user's machine. If they have different security settings (maybe some of them told gatekeeper to go away and others didn't), maybe that'd explain the behavior difference?

I don't have any evidence to back up any of those guesses, FWIW. Semi-educated guesses are the best I have to offer right now. You've definitely caught my interest though :)

dtruss (which is I guess the macOS strace) of one of those stalled processes would probably tell you a lot.

**Reply by justinpapezsonos** [reply]
> Which r21? If you go back far enough (r21b or older) the binaries won't be signed/notarized. If gatekeeper is for some reason blocking all those processes only when they're signed/notarized, that's a difference. I am inclined to blame gatekeeper here, since this does look very similar to the behavior you'd find on Windows when a process is blocked waiting on AV.
> 
> Are all your developers downloading the NDK the same way? Via the SDK manager, or via direct download? For direct downloads there are two different NDKs available for Mac: an app bundle (Apple requires the NDK be distributed this way for notarization, which gatekeeper requires) and the normal zip file, which cannot be notarized. The SDK can only distribute the zip file, but because the SDK manager doesn't quarantine its downloads those binaries don't require notarization. Maybe the behavior split is based on the download type/method?
> 
> Another possible split would be the settings on each user's machine. If they have different security settings (maybe some of them told gatekeeper to go away and others didn't), maybe that'd explain the behavior difference?
> 
> I don't have any evidence to back up any of those guesses, FWIW. Semi-educated guesses are the best I have to offer right now. You've definitely caught my interest though :)
> 
> dtruss (which is I guess the macOS strace) of one of those stalled processes would probably tell you a lot.

NDK r21 version 21.0.6113669. All developers are downloading the NDK the same way via the SDK manager. Specifically, the NDK version and installation is controlled via our build environment. We define the version of the NDK in the app level build.gradle file and then parse the version of the NDK and install it via CMake.

app/build.gradle
```
ndkVersion "21.0.6113669"
```

android_toolchain.cmake
```
string(REGEX REPLACE "^[ \t]*ndkVersion[ \t]*\"(.*)\".*$" "\\1" ACR_NDK_VER ${NDK_VERSION_STRING})
...
set(ACR_NDK_PKG "ndk;${ACR_NDK_VER}")
...
set(ANDROID_SDKMANAGER $ENV{ANDROID_SDK_ROOT}/cmdline-tools/latest/bin/sdkmanager)
...
execute_process(
  COMMAND ${ANDROID_SDKMANAGER} "${ACR_NDK_PKG}"
  RESULT_VARIABLE INSTALLED_PKG
  # OUTPUT_VARIABLE INSTALL_OUTPUT
  # ERROR_VARIABLE INSTALL_ERROR
)
```
Going to circle back with the developers encountering the issues and chat with them about the splits you've pointed out. Will post an update again in a bit :). Thanks for the continued help!

**Reply by justinpapezsonos** [reply]
> Which r21? If you go back far enough (r21b or older) the binaries won't be signed/notarized. If gatekeeper is for some reason blocking all those processes only when they're signed/notarized, that's a difference. I am inclined to blame gatekeeper here, since this does look very similar to the behavior you'd find on Windows when a process is blocked waiting on AV.
> ...
> Another possible split would be the settings on each user's machine. If they have different security settings (maybe some of them told gatekeeper to go away and others didn't), maybe that'd explain the behavior difference?

I had a fellow developer who is experiencing exuberant build times disable gatekeeper via `sudo spctl --master-disable`. Sadly, re-building with gatekeeper disabled had no impact on performance time.

**Reply by DanAlbert** [reply]
That's the original r21 release, so it is pre-signing. You're missing everything described in r21b, c, d, and e: https://github.com/android/ndk/wiki/Changelog-r21

Would be interesting if upgrading to a newer r21 causes the same issue.

**Reply by justinpapezsonos** [reply]
> That's the original r21 release, so it is pre-signing. You're missing everything described in r21b, c, d, and e: https://github.com/android/ndk/wiki/Changelog-r21
> 
> Would be interesting if upgrading to a newer r21 causes the same issue.

As you mentioned, we have not yet tried compiling against a later version of NDK r21 (such as b, c, d, or e). However, it may be worth mentioning that we have compiled against NDK r22, and none of our developers experienced increased build times.

**Reply by justinpapezsonos** [reply]
Just wanted to close the loop, your hunch was right @DanAlbert, the root issue is our corporate AV program. Thanks again for all the responses & help!

**Reply by DanAlbert** [reply]
Great, thanks for following up. I edited Pirama's initial comment a bit to make the thread more readable (seems only the first comment in a thread will be attached to the question as the "answer" when selected).
